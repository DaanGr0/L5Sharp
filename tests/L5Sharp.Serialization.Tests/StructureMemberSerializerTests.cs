using System;
using System.Collections.Generic;
using System.Xml.Linq;
using ApprovalTests;
using ApprovalTests.Reporters;
using FluentAssertions;
using L5Sharp.Core;
using L5Sharp.Creators;
using L5Sharp.Enums;
using L5Sharp.Types;
using NUnit.Framework;

namespace L5Sharp.Serialization.Tests
{
    [TestFixture]
    public class StructureMemberSerializerTests
    {
        private StructureMemberSerializer _serializer;

        [SetUp]
        public void Setup()
        {
            _serializer = new StructureMemberSerializer();
        }
        
        [Test]
        public void Serialize_Null_ShouldThrowArgumentNullException()
        {
            FluentActions.Invoking(() => _serializer.Serialize(null!)).Should().Throw<ArgumentException>();
        }
        
        [Test]
        public void Serialize_WhenCalled_ShouldNotBeNull()
        {
            var component = new Member<IDataType>("Test", new StructureType("Test"));

            var xml = _serializer.Serialize(component);

            xml.Should().NotBeNull();
        }
        
        [Test]
        [UseReporter(typeof(DiffReporter))]
        public void Serialize_DataValueValueMembers_ShouldBeApproved()
        {
            var type = new StructureType("Test", new List<IMember<IDataType>>
            {
                Member.Create<BOOL>("BoolMember"),
                Member.Create<SINT>("SintMember"),
                Member.Create<INT>("IntMember"),
                Member.Create<DINT>("DintMember"),
                Member.Create<LINT>("LintMember"),
                Member.Create<REAL>("RealMember")
            });

            var component = new Member<IDataType>("Test", type);

            var xml = _serializer.Serialize(component);

            Approvals.VerifyXml(xml.ToString());
        }
        
        [Test]
        [UseReporter(typeof(DiffReporter))]
        public void SerializeArrayMembers_ShouldBeApproved()
        {
            var type = new StructureType("Test", new List<IMember<IDataType>>
            {
                Member.Create<DINT>("SimpleMember", new Dimensions(10)),
                Member.Create<TIMER>("ComplexMember", 5),
                Member.Create<STRING>("StringMember", 2)
            });
            
            var component = new Member<IDataType>("Test", type);

            var xml = _serializer.Serialize(component);

            Approvals.VerifyXml(xml.ToString());
        }
        
        [Test]
        [UseReporter(typeof(DiffReporter))]
        public void Serialize_StructureMembers_ShouldBeApproved()
        {
            var type = new StructureType("Test", new List<IMember<IDataType>>
            {
                Member.Create<STRING>("StringMember"),
                Member.Create<TIMER>("SintMember"),
                Member.Create<COUNTER>("IntMember"),
            });
            
            var component = new Member<IDataType>("Test", type);

            var xml = _serializer.Serialize(component);

            Approvals.VerifyXml(xml.ToString());
        }
        
        [Test]
        public void Deserialize_Null_ShouldThrowArgumentNullException()
        {
            FluentActions.Invoking(() => _serializer.Deserialize(null!)).Should().Throw<ArgumentException>();
        }

        [Test]
        public void Deserialize_InvalidElementName_ShouldThrowArgumentException()
        {
            const string xml = @"<Invalid></Invalid>";
            var element = XElement.Parse(xml);

            FluentActions.Invoking(() => _serializer.Deserialize(element)).Should().Throw<ArgumentException>()
                .WithMessage($"Element 'Invalid' not valid for the serializer {_serializer.GetType()}.");
        }

        [Test]
        public void Deserialize_SimpleStructureMember_ShouldNotBeNull()
        {
            var xml = XElement.Parse(GetSimpleStructureMember());

            var component = _serializer.Deserialize(xml);

            component.Should().NotBeNull();
        }
        
        [Test]
        public void Deserialize_SimpleStructureMember_ShouldHaveExpectedProperties()
        {
            var xml = XElement.Parse(GetSimpleStructureMember());

            var component = _serializer.Deserialize(xml);

            component.Name.Should().Be("SimpleMember");
            component.DataType.Name.Should().Be("SimpleType");
            component.DataType.Should().BeOfType<StructureType>();
            component.Dimensions.Should().Be(Dimensions.Empty);
            component.Radix.Should().Be(Radix.Null);
            component.Description.Should().BeEmpty();
            component.MemberType.Should().Be(MemberType.StructureMember);
            component.DataType.As<IComplexType>().Members.Should().HaveCount(5);
        }

        private static string GetSimpleStructureMember()
        {
            return  @"<StructureMember Name=""SimpleMember"" DataType=""SimpleType"">
                <DataValueMember Name=""BoolMember"" DataType=""BOOL"" Value=""0""/>
                <DataValueMember Name=""SintMember"" DataType=""SINT"" Radix=""Hex"" Value=""16#00""/>
                <DataValueMember Name=""IntMember"" DataType=""INT"" Radix=""Octal"" Value=""8#000_000""/>
                <DataValueMember Name=""DintMember"" DataType=""DINT"" Radix=""ASCII"" Value=""'$00$00$00$00'""/>
                <DataValueMember Name=""LintMember"" DataType=""LINT"" Radix=""Decimal"" Value=""0""/>
                </StructureMember>";
        }
    }
}