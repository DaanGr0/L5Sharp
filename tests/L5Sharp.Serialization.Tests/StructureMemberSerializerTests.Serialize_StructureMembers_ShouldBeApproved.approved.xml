<StructureMember Name="Test" DataType="Test">
  <StructureMember Name="StringMember" DataType="STRING">
    <DataValueMember Name="LEN" DataType="DINT" Radix="Decimal" Value="82" />
    <ArrayMember Name="DATA" DataType="SINT" Dimensions="82" Radix="Ascii">
      <Element Index="[0]" Value="'$00'" />
      <Element Index="[1]" Value="'$00'" />
      <Element Index="[2]" Value="'$00'" />
      <Element Index="[3]" Value="'$00'" />
      <Element Index="[4]" Value="'$00'" />
      <Element Index="[5]" Value="'$00'" />
      <Element Index="[6]" Value="'$00'" />
      <Element Index="[7]" Value="'$00'" />
      <Element Index="[8]" Value="'$00'" />
      <Element Index="[9]" Value="'$00'" />
      <Element Index="[10]" Value="'$00'" />
      <Element Index="[11]" Value="'$00'" />
      <Element Index="[12]" Value="'$00'" />
      <Element Index="[13]" Value="'$00'" />
      <Element Index="[14]" Value="'$00'" />
      <Element Index="[15]" Value="'$00'" />
      <Element Index="[16]" Value="'$00'" />
      <Element Index="[17]" Value="'$00'" />
      <Element Index="[18]" Value="'$00'" />
      <Element Index="[19]" Value="'$00'" />
      <Element Index="[20]" Value="'$00'" />
      <Element Index="[21]" Value="'$00'" />
      <Element Index="[22]" Value="'$00'" />
      <Element Index="[23]" Value="'$00'" />
      <Element Index="[24]" Value="'$00'" />
      <Element Index="[25]" Value="'$00'" />
      <Element Index="[26]" Value="'$00'" />
      <Element Index="[27]" Value="'$00'" />
      <Element Index="[28]" Value="'$00'" />
      <Element Index="[29]" Value="'$00'" />
      <Element Index="[30]" Value="'$00'" />
      <Element Index="[31]" Value="'$00'" />
      <Element Index="[32]" Value="'$00'" />
      <Element Index="[33]" Value="'$00'" />
      <Element Index="[34]" Value="'$00'" />
      <Element Index="[35]" Value="'$00'" />
      <Element Index="[36]" Value="'$00'" />
      <Element Index="[37]" Value="'$00'" />
      <Element Index="[38]" Value="'$00'" />
      <Element Index="[39]" Value="'$00'" />
      <Element Index="[40]" Value="'$00'" />
      <Element Index="[41]" Value="'$00'" />
      <Element Index="[42]" Value="'$00'" />
      <Element Index="[43]" Value="'$00'" />
      <Element Index="[44]" Value="'$00'" />
      <Element Index="[45]" Value="'$00'" />
      <Element Index="[46]" Value="'$00'" />
      <Element Index="[47]" Value="'$00'" />
      <Element Index="[48]" Value="'$00'" />
      <Element Index="[49]" Value="'$00'" />
      <Element Index="[50]" Value="'$00'" />
      <Element Index="[51]" Value="'$00'" />
      <Element Index="[52]" Value="'$00'" />
      <Element Index="[53]" Value="'$00'" />
      <Element Index="[54]" Value="'$00'" />
      <Element Index="[55]" Value="'$00'" />
      <Element Index="[56]" Value="'$00'" />
      <Element Index="[57]" Value="'$00'" />
      <Element Index="[58]" Value="'$00'" />
      <Element Index="[59]" Value="'$00'" />
      <Element Index="[60]" Value="'$00'" />
      <Element Index="[61]" Value="'$00'" />
      <Element Index="[62]" Value="'$00'" />
      <Element Index="[63]" Value="'$00'" />
      <Element Index="[64]" Value="'$00'" />
      <Element Index="[65]" Value="'$00'" />
      <Element Index="[66]" Value="'$00'" />
      <Element Index="[67]" Value="'$00'" />
      <Element Index="[68]" Value="'$00'" />
      <Element Index="[69]" Value="'$00'" />
      <Element Index="[70]" Value="'$00'" />
      <Element Index="[71]" Value="'$00'" />
      <Element Index="[72]" Value="'$00'" />
      <Element Index="[73]" Value="'$00'" />
      <Element Index="[74]" Value="'$00'" />
      <Element Index="[75]" Value="'$00'" />
      <Element Index="[76]" Value="'$00'" />
      <Element Index="[77]" Value="'$00'" />
      <Element Index="[78]" Value="'$00'" />
      <Element Index="[79]" Value="'$00'" />
      <Element Index="[80]" Value="'$00'" />
      <Element Index="[81]" Value="'$00'" />
    </ArrayMember>
  </StructureMember>
  <StructureMember Name="SintMember" DataType="TIMER">
    <DataValueMember Name="PRE" DataType="DINT" Radix="Decimal" Value="0" />
    <DataValueMember Name="ACC" DataType="DINT" Radix="Decimal" Value="0" />
    <DataValueMember Name="EN" DataType="BOOL" Value="0" />
    <DataValueMember Name="TT" DataType="BOOL" Value="0" />
    <DataValueMember Name="DN" DataType="BOOL" Value="0" />
  </StructureMember>
  <StructureMember Name="IntMember" DataType="COUNTER">
    <DataValueMember Name="PRE" DataType="DINT" Radix="Decimal" Value="0" />
    <DataValueMember Name="ACC" DataType="DINT" Radix="Decimal" Value="0" />
    <DataValueMember Name="CU" DataType="BOOL" Value="0" />
    <DataValueMember Name="CD" DataType="BOOL" Value="0" />
    <DataValueMember Name="DN" DataType="BOOL" Value="0" />
    <DataValueMember Name="OV" DataType="BOOL" Value="0" />
    <DataValueMember Name="UN" DataType="BOOL" Value="0" />
  </StructureMember>
</StructureMember>